### ISSUE TITLE
DB50XG emulation or S-YXG50 VSTi support

### ISSUE THREAD

**Issue by Nazosan** [joncampbell123/dosbox-x#3632]
### What you want

A mechanism of having some sort of built in support for Yamaha XG MIDI synthesis -- preferably something not platform-specific.

### Describe alternatives you've considered

So far I've been relying on a VSTi MIDI driver using the old (32-bit x86 only) S-YXG50 VSTi plugin and setting DOSBox/DOSBox-X to use the Windows MIDI mapper instead of its own internal synthesis options.  The success rate of this is getting lower and lower every year however.  Microsoft has been slowly phasing out MIDI support in Windows in various ways with first the removal of any way to get to the MIDI mapper's configuration, then a removal of the MIDI mapper itself, and then recently in a Windows 10 update they changed a bunch of stuff in how it actually does the settings in the registry, breaking compatibility with existing solutions.  Third party solutions are having varying success in dealing with these issues and the development of the current VSTi MIDI driver seems to have completely halted too with unofficial third party builds being basically the best solution left currently.  Eventually even that may no longer work.  Someday Microsoft will likely completely break any ability to use any MIDI driver other than their own (and may even someday remove that.)  Eventually third parties will stop working to fix the issues they create in this regard.  What's more, Linux users probably have an even harder time doing anything for this.  Eventually I think we'll need to rely solely on built-in solutions for things like legacy machine emulation and while it obviously can't emulate every MIDI device ever, I think the DB50XG would be a great addition to those things it does.

### Additional information

I realize this is a fairly tough request and not likely to be considered right away probably, but I want to at least put it in people's minds to think about it.  The Yamaha DB50XG daughterboard card was one of the absolute best in MIDI synthesizers of its time for many things.  While not as advanced in regards to DLS (SoundFonts/etc) as the Creative "AWE" series that we're already essentially sort of emulating -- even improving upon -- with the SoundFont libraries already built into DOSBox/DOSBox-X, it did have its own advantages.  Of course the biggest advantage is that it supported not only General MIDI, but also further extensions in Yamaha's own XG standard.  This can be less relevant to gaming in most cases as admittedly not a lot of games truly took advantage of this, but there are some MIDI players, editors, etc some people apparently still use that are having less and less positive results with modern systems.  However, there are many people who might actually be glad to see such a thing in Windows tools, some of which they actually still use today apparently.

For gaming though, I find that the DB50XG and S-YXG50 by extension (Yamaha essentially wrote S-YXG50 as a software implementation of more or less the exact functionality of the DB50XG card) has a somewhat more unexpected advantage in that its sound seems to generically suit the largest number of games as just a generic default MIDI device compared to any soundfont I've ever been able to find so far.  (To be clear, soundfonts can sound better, but generally they're more "personal" in that A. they're very much down to personal preference and B. they tend to suit only a few games -- or even one -- and sometimes even don't do well on one track versus another even within the games they work well with.)  

The simplest implementation would probably be a VSTi host.  There are a lot of open source implementations of this I believe and it probably wouldn't be terribly hard to integrate.  However, while I don't know much about how it actually works, it looks like VSTi plugins are essentially binaries that have to be compatible with the host system.  That means a plugin intended for Windows probably won't work in Linux and vice versa and none of the ones made for normal PCs will work on other platforms (say ARM for instance) making such a system probably largely useless for ports outside of x86 PCs.  (Also, technically if we ever move on to 128-bit processors that would make 32-bit plugins stop working I guess, but I'll admit this isn't likely to happen anywhere remotely soon.)

A more universal solution would be an implementation of some sort of XG capable emulated device, perhaps still having the user supply the sample ROM separately (it seems the 4MiB ROM is available on the Web seemingly freely, but I'm still not clear on if that is 100% legal or not.  Certainly it couldn't be included with DOSBox-X, but it wouldn't exactly be hard for users to get it or to get the 2MiB sample from the S-YXG50 software itself.)  It may not necessarily have to exactly emulate a DB50XG daughterboard as it's largely just a matter of supporting a wavetable card at all and then implementing support for the sample ROM and the XG synthesis functionality.  (I don't think it had any particularly specialized specialties to the hardware in regards to stuff that would have to be implemented for games and software to work with it, though Windows drivers might want to identify it.)

While this probably won't happen any time soon, I just wanted to at least submit it for consideration.

### Have you checked that no similar feature request(s) exist?

- [X] I have searched and didn't find any similar feature request.

### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

**Comment by sev7en2507** [comment]
Hello, I am interested to. Thanks

**Comment by Torinde** [comment]
Related request to emulate the Yamaha MU80: https://github.com/dosbox-staging/dosbox-staging/pull/2326#issuecomment-1464493925

**Comment by ThomasEricB** [comment]
> Related request to emulate the Yamaha MU80: [dosbox-staging/dosbox-staging#2326 (comment)](https://github.com/dosbox-staging/dosbox-staging/pull/2326#issuecomment-1464493925)

My PR has nothing to do with the Yamaha MU80! LOL. Though, it IS something I would like to see and I've asked @johnnovak if it was possible

**Comment by ThomasEricB** [comment]
> Related request to emulate the Yamaha MU80: [dosbox-staging/dosbox-staging#2326 (comment)](https://github.com/dosbox-staging/dosbox-staging/pull/2326#issuecomment-1464493925)

I 100% stand behind this idea! @johnnovak make it happen!

**Comment by Torinde** [comment]
In addition to XG: I see MIDI has other extensions such as GM2, DLS, etc.
In addition to VSTi: I see [DOSbox Core](https://github.com/realnc/dosbox-core) supports SFZ via [BASSMIDI](https://www.un4seen.com/)

**Comment by nickkeane** [comment]
EmuSC (Roland SC-55 and variants) is probably the closest we're going to get to a "Modern" emulated MIDI device, if and when that ever gets finished and integrated.

In the meantime, Fluidsynth soundfonts are probably the best way imitate a General MIDI device, outside of maintaining an External MIDI or VSTi. However, the .sf2 format is old and limited in what it can imitate. There are proposals for new standards like .sfz and .sf3 that address many of these problems, but they haven't seen wide adoption (yet).

https://github.com/davy7125/soundfont-standard-v3

I have ROM dumps for the MU50, MU80, and MU100, but I don't know if they're complete sets of the IC dumps that are needed to fully emulate those devices. If someone with a lot of knowledge of MIDI audio processing and MIDI ROM reverse-engineering wants to take a crack at it, I can be hit up for the roms.

**Comment by johnnovak** [comment]
Guys, I like your enthusiasm, but you don't really grasp the difficulty of creating an accurate S-YXG50 emulation, given it's not documented anywhere how it really works. That's a multi-year effort at minimum, and even then it's questionable how accurate you can make it. Incorporating Yamaha's original code into the project from the S-YXG50 as a binary blob is questionable too... Plus you'd need to stub/mock Windows calls, and possibly run the code on an emulated Intel CPU. OR you'd fully reverse engineer it, and attempt to reimplement in C. All of these are very difficult to do, and the legality is questionable to varying degrees with all approaches.

Just use SAVIHost on Windows to host the S-YXG50 VSTi, then connect it to DOSBOx with loopMIDI as an external MIDI device. Instructions in this video, and I'll create a wiki article about this sometime. Many people can run SAVIHost or something similar with WINE under Linux just fine.

> However, the .sf2 format is old and limited in what it can imitate. 

Agreed, forget about Fluidsynth / SF2 for this.

https://www.youtube.com/watch?v=WumRboSfn90

> I 100% stand behind this idea! @johnnovak make it happen!

Sorry, I won't spend time on this; see my tips above how to get it working here and now 😄 

However, if anyone will start working on this, I'll be surely watching 🍿 😎

**Comment by johnnovak** [comment]
> I have ROM dumps for the MU50, MU80, and MU100, but I don't know if they're complete sets of the IC dumps that are needed to fully emulate those devices. If someone with a lot of knowledge of MIDI audio processing and MIDI ROM reverse-engineering wants to take a crack at it, I can be hit up for the roms.

Well, as I said I won't be working on this, but those ROM dumps are still worth preserving, just in case. How about uploading them to archive.org or the ROM dump library at vogons.org?

**Comment by johnnovak** [comment]
By the way @Nazosan, not sure if you were inspired by my blog post or PhilsComputerLabs recent videos to raise this ticket (seems too much of a coincide if you weren't 😏 ), but here's my comprehensive comparison of the hardware MU80, the S-YXG50 and both the hardware SC-55 and the SCVA with metric tons of lossless audio recordings:

https://blog.johnnovak.net/2023/03/05/grand-ms-dos-gaming-general-midi-showdown/

**Comment by Nazosan** [comment]
No.  I've been using Yamaha XG for MIDI playback since I first got a YMF-724 based card but really it was more of a "trying to help people get set up with DOSBox with MIDI that I at least personally feel sounds better in specific games without 5000 steps" kind of thing.  I'd really love to at least see something more encapsulated where at least I could rig up something in a semi-portable format without them having to install or configure anything.  But also I had kind of wished such a project might someday happen in regards to how it could also apply to other software and platforms (for example, say maybe ScummVM on Android which obviously isn't going to be running any VSTis for x86 systems.)

**Comment by johnnovak** [comment]
Makes sense I also wish an emulation of the MU-series modules was available, and I agree it would be nice to have a cross-platform solution. It's just... we don't have it, and the amount of work required to do (properly) it is highly non-trivial. So yeah, for right now, the VSTi is the answer. But that would "bit-rot" away in 10-20 etc. years from now, I get that angle too... Well, I guess that's even more of an issue with Roland's payware SCVA plugin.
